src/box/space.c

```diff
@@ -403,6 +457,7 @@ space_delete(struct space *space)
 		space_cleanup_constraints(space);
 		tuple_format_unref(space->format);
 	}
+	space_unpin_collations(space);
```

Comment:
Can't we replace this with an assertion?
```C
assert(rlist_empty(&space->coll_id_holders));
```

Reply:
No, `space_delete()` must unpin them. Here is the testcase:
```lua
local coll_name = 'my coll'
box.internal.collation.create(coll_name, 'ICU', 'ru-RU', {})
print('--------- space create ---------')
local s = box.schema.space.create('test',
    {format = {{name = 'p'}, {name = 's', type = 'string', collation = coll_name}}})
print('--------- index create ---------')
s:create_index('pk')
print('--------- box.begin ---------')
box.begin()
print('--------- space drop ---------')
box.space.test:drop()
print('--------- collation drop ---------')
box.internal.collation.drop(coll_name)
print('--------- box.rollback ---------')
box.rollback()
```

My patch does 5 pins and 5 unpins:
```lua
--------- space create ---------
on_replace_dd_space()
space_create()
↳ space_pin_collations(test)
--------- index create ---------
alter_space_do()
space_create()
↳ space_pin_collations(test)
↳ space_unpin_collations(test)
space_delete(test)
↳ space_unpin_collations(test)
--------- box.begin ---------
--------- space drop ---------
alter_space_do()
space_create()
↳ space_pin_collations(test)
↳ space_unpin_collations(test)
on_replace_dd_space()
↳ space_unpin_collations(test)
--------- collation drop ---------
--------- box.rollback ---------
on_drop_space_rollback()
↳ space_pin_collations(test)
alter_space_rollback()
↳ space_pin_collations(test)
space_delete(test)
↳ space_unpin_collations(test)
```

Assert without unpin in `space_delete()` (5 pins, 3 unpins):
```lua
--------- space create ---------
on_replace_dd_space()
space_create()
↳ space_pin_collations(test)
--------- index create ---------
alter_space_do()
space_create()
↳ space_pin_collations(test)
↳ space_unpin_collations(test)
space_delete(test)
--------- box.begin ---------
--------- space drop ---------
alter_space_do()
space_create()
↳ space_pin_collations(test)
↳ space_unpin_collations(test)
on_replace_dd_space()
↳ space_unpin_collations(test)
--------- collation drop ---------
--------- box.rollback ---------
on_drop_space_rollback()
↳ space_pin_collations(test)
alter_space_rollback()
↳ space_pin_collations(test)
space_delete(test)
tarantool: ./src/box/space.c:449: space_delete: Assertion `rlist_empty(&space->coll_id_holders)' failed.
```